PROJECT_MCPU/MCPU/DOC/GENERATED/man/man3/PCB302.3
.TH "PCB302" 3 "MCPU" \" -*- nroff -*-
.ad l
.nh
.SH NAME
PCB302
.SH SYNOPSIS
.br
.PP
.PP
\fR#include <PCB302\&.h>\fP
.PP
Inherits \fBCanDeviceProtocol\fP\&.
.SS "Classes"

.in +1c
.ti -1c
.RI "class \fBProtocolStructure\fP"
.br
.RI "This class implement the protocol data structure as described in the protocol specification\&. "
.in -1c
.SS "Public Types"

.in +1c
.ti -1c
.RI "enum class \fBPatientProtection\fP { \fBUNDETECTED\fP = 0, \fBPOSITIONED\fP, \fBSHIFTED\fP }"
.br
.RI "This is the enumeration of the possible Patient protection presence\&. "
.ti -1c
.RI "enum class \fBpaddleCodes\fP { \fBPADDLE_PROSTHESIS\fP =0, \fBPADDLE_BIOP2D\fP, \fBPADDLE_BIOP3D\fP, \fBPADDLE_TOMO\fP, \fBPADDLE_24x30_CONTACT\fP, \fBPADDLE_18x24_C_CONTACT\fP, \fBPADDLE_18x24_L_CONTACT\fP, \fBPADDLE_18x24_R_CONTACT\fP, \fBPADDLE_10x24_CONTACT\fP, \fBPADDLE_9x21_MAG\fP, \fBPADDLE_9x9_MAG\fP, \fBPADDLE_D75_MAG\fP, \fBPADDLE_LEN\fP, \fBPADDLE_NOT_DETECTED\fP = PADDLE_LEN }"
.br
.RI "This enumeration class defines the Paddle identified by the Application\&. "
.in -1c

Public Types inherited from \fBCanDeviceProtocol\fP
.in +1c
.ti -1c
.RI "enum class \fBCommandRegisterErrors\fP { \fBCommandRegisterErrors::COMMAND_NO_ERROR\fP = 0, \fBCommandRegisterErrors::COMMAND_ERROR_BUSY\fP, \fBCommandRegisterErrors::COMMAND_ERROR_INVALID_PARAM\fP, \fBCommandRegisterErrors::COMMAND_ERROR_MOMENTARY_DISABLED\fP, \fBCommandRegisterErrors::COMMAND_INVALID_DEVICE\fP =253, \fBCommandRegisterErrors::COMMAND_COMMUNICATION_ERROR\fP =254, \fBCommandRegisterErrors::COMMAND_DEVICE_TMO\fP = 255 }"
.br
.RI "This is the enumeration of the possible command executed errors\&. "
.ti -1c
.RI "enum class \fBstatus_options\fP { \fBstatus_options::WAITING_CAN_DRIVER_CONNECTION\fP = 0, \fBstatus_options::WAITING_REVISION\fP, \fBstatus_options::DEVICE_CONFIGURATION\fP, \fBstatus_options::DEVICE_RUNNING\fP, \fBstatus_options::DEVICE_SIMULATOR\fP, \fBstatus_options::LEN\fP, \fBstatus_options::UNDEF\fP = LEN }"
.br
.RI "This is the enumeration of the Module's internal operating status\&.
.br
See the \fBgetModuleStatus()\fP method\&. "
.ti -1c
.RI "enum class \fBbootloader_options\fP { \fBbootloader_options::BOOTLOADER_NOT_PRESENT\fP = 0, \fBbootloader_options::BOOTLOADER_RUNNING\fP = 1, \fBbootloader_options::BOOTLOADER_PRESENT\fP = 2, \fBbootloader_options::BOOTLOADER_UNCKNOWN_STAT\fP }"
.br
.RI "This is the enumeration calss defining the current status of the remote bootloader activity\&. "
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBPCB302\fP ()"
.br
.in -1c

Public Member Functions inherited from \fBCanDeviceProtocol\fP
.in +1c
.ti -1c
.RI "\fBCanDeviceProtocol\fP (unsigned char devid, LPCWSTR devname)"
.br
.ti -1c
.RI "void \fBrunMode\fP (void)"
.br
.RI "This function activate the Module in Real mode\&. "
.ti -1c
.RI "void \fBsimulMode\fP (void)"
.br
.RI "This function activate the Module in Simulation mode\&. "
.ti -1c
.RI "bool \fBisSimulatorMode\fP (void)"
.br
.RI "This function tests if the module is running in simulation mode\&. "
.ti -1c
.RI "bool \fBisCommunicationError\fP (void)"
.br
.RI "This function tests if the module is in communication error condition\&. "
.ti -1c
.RI "\fBstatus_options\fP \fBgetModuleStatus\fP (void)"
.br
.RI "This function returns the Module operating status\&. "
.ti -1c
.RI "\fBbootloader_options\fP \fBgetBootStatus\fP (void)"
.br
.RI "This function returns the current status of the bootloader activities on the remote device\&. "
.ti -1c
.RI "System::String ^ \fBgetBootRevision\fP (void)"
.br
.RI "This function returns a string with the current detected Bootloader revision code\&. "
.ti -1c
.RI "System::String ^ \fBgetAppRevision\fP (void)"
.br
.RI "This function returns a string with the current detected remote firmware revision code\&. "
.ti -1c
.RI "bool \fBcommandNoWaitCompletion\fP (unsigned char code, unsigned char d0, unsigned char d1, unsigned char d2, unsigned char d3, int tmo)"
.br
.RI "This is a non blocking thread safe function sending a command to the remote device\&. "
.ti -1c
.RI "bool \fBcommandNoWaitCompletion\fP (\fBCanDeviceCommand\fP^ command, int tmo)"
.br
.ti -1c
.RI "\fBCanDeviceCommandResult\fP ^ \fBcommandWaitCompletion\fP (unsigned char code, unsigned char d0, unsigned char d1, unsigned char d2, unsigned char d3, int tmo, Object^ src)"
.br
.RI "This is a blocking thread safe function sending a command to the remote device\&. "
.ti -1c
.RI "\fBCanDeviceCommandResult\fP ^ \fBcommandWaitCompletion\fP (\fBCanDeviceCommand\fP^ command, int tmo, Object^ src)"
.br
.ti -1c
.RI "bool \fBisCommandCompleted\fP (void)"
.br
.RI "This function returns the current status of the pending command\&. "
.ti -1c
.RI "bool \fBisCommandError\fP (void)"
.br
.RI "This function returns true if the command is terminated with an error condition\&. "
.ti -1c
.RI "unsigned char \fBgetCommandResult0\fP (void)"
.br
.RI "In case of command completed without errors, this function returns the command returned data-0\&. "
.ti -1c
.RI "unsigned char \fBgetCommandResult1\fP (void)"
.br
.RI "In case of command completed without errors, this function returns the command returned data-1\&. "
.ti -1c
.RI "\fBCommandRegisterErrors\fP \fBgetCommandError\fP (void)"
.br
.RI "This function returns the last error condition of an executed command\&. "
.ti -1c
.RI "System::String ^ \fBgetCanCommunicationMonitorString\fP (void)"
.br
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static System::String ^ \fBgetPaddleName\fP (\fBpaddleCodes\fP paddle)"
.br
.ti -1c
.RI "static unsigned short \fBgetRawPosition\fP (void)"
.br
.ti -1c
.RI "static unsigned short \fBgetRawForce\fP (void)"
.br
.ti -1c
.RI "static unsigned short \fBgetPaddlePosition\fP (void)"
.br
.ti -1c
.RI "static unsigned short \fBgetPaddleForce\fP (void)"
.br
.ti -1c
.RI "static unsigned short \fBgetThickness\fP (void)"
.br
.ti -1c
.RI "static unsigned short \fBgetForce\fP (void)"
.br
.RI "This function returnrs the current thickness in mm\&. "
.ti -1c
.RI "static bool \fBisDownwardActivationStatus\fP (void)"
.br
.RI "This function returnrs the current compression force in N\&. "
.ti -1c
.RI "static bool \fBisUpwardActivationStatus\fP (void)"
.br
.ti -1c
.RI "static bool \fBisIdleStatus\fP (void)"
.br
.ti -1c
.RI "static void \fBsetCompressorUnlock\fP (void)"
.br
.RI "This function unlocks the compression\&. "
.ti -1c
.RI "static \fBpaddleCodes\fP \fBgetPaddleCode\fP (System::String^ tag)"
.br
.RI "This function returns the paddle code from the paddle name 
.br
 "
.ti -1c
.RI "static \fBpaddleCodes\fP \fBgetDetectedPaddleCode\fP (void)"
.br
.RI "This function returns the current detected paddle code\&. "
.ti -1c
.RI "static System::Byte \fBgetPaddleCollimationFormat\fP (\fBpaddleCodes\fP paddle_code)"
.br
.RI "This function returns the index of the collimation format associated at the paddle\&. "
.ti -1c
.RI "static System::Byte \fBgetDetectedPaddleCollimationFormat\fP (void)"
.br
.RI "This function returns the collimation format index associated to the detected paddle\&. "
.ti -1c
.RI "static \fBPatientProtection\fP \fBgetPatientProtection\fP (void)"
.br
.RI "This function returns the current status of the Patient Protection presence\&. "
.ti -1c
.RI "static unsigned char \fBgetMagnifierFactor\fP (void)"
.br
.RI "This function return the current detected Magnifier Factor\&. "
.in -1c
.SS "Static Public Attributes"

.in +1c
.ti -1c
.RI "static \fBPCB302\fP ^ \fBdevice\fP = gcnew \fBPCB302\fP()"
.br
.ti -1c
.RI "static const cli::array< System::String^> ^ \fBpaddle_names\fP"
.br
.in -1c
.SS "Protected Member Functions"

.in +1c
.ti -1c
.RI "void \fBrunningLoop\fP (void) override"
.br
.RI "This routine shall be overridden by the subclass module in order to inplement the specific module features\&. "
.ti -1c
.RI "bool \fBconfigurationLoop\fP (void) override"
.br
.in -1c

Protected Member Functions inherited from \fBCanDeviceProtocol\fP
.in +1c
.ti -1c
.RI "\fBRegister\fP ^ \fBreadCommandRegister\fP (void)"
.br
.RI "This function executes the read of the special device command register\&. "
.ti -1c
.RI "\fBRegister\fP ^ \fBreadErrorRegister\fP (void)"
.br
.RI "This function executes the read of the special device error register\&. "
.ti -1c
.RI "\fBRegister\fP ^ \fBreadStatusRegister\fP (unsigned char index)"
.br
.RI "This function executes the read of the STATUS register of given INDEX\&. "
.ti -1c
.RI "\fBRegister\fP ^ \fBreadDataRegister\fP (unsigned char index)"
.br
.RI "This function executes the read of the DATA register of given INDEX\&. "
.ti -1c
.RI "bool \fBwriteParamRegister\fP (unsigned char index, unsigned char d0, unsigned char d1, unsigned char d2, unsigned char d3)"
.br
.RI "This function executes the write of the PARAMETER(index)\&. "
.ti -1c
.RI "bool \fBwriteParamRegister\fP (unsigned char idx, \fBRegister\fP^ reg)"
.br
.RI "This function executes the write of the PARAMETER(index)\&. "
.ti -1c
.RI "bool \fBwriteDataRegister\fP (unsigned char index, unsigned char d0, unsigned char d1, unsigned char d2, unsigned char d3)"
.br

PROJECT_MCPU/MCPU/MCPU_MASTER/DEVICES/PCB302.cpp

#include "PCB303.h"
#include "PCB302.h"
#include "ConfigurationFiles.h"
#include "../gantry_global_status.h"
#include "ArmMotor.h"
#include "Notify.h"
#include <thread>


void PCB302::moduleInitialize(void) {

	// Checks if the position has been already calibrated	
	if (System::Convert::ToByte(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_COMPRESSOR)[PaddleConfig::COMPRESSOR_CALIBRATION_POSITION_STATUS]) == 0) {
		Notify::activate(Notify::messages::WARNING_POSITION_NOT_CALIBRATED);
		position_calibrated = false;
	}
	else {
		position_calibrated = true;
		Notify::deactivate(Notify::messages::WARNING_POSITION_NOT_CALIBRATED);
	}

	// Checks if the force has been already calibrated	
	if (System::Convert::ToByte(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_COMPRESSOR)[PaddleConfig::COMPRESSOR_CALIBRATION_FORCE_STATUS]) == 0) {
		Notify::activate(Notify::messages::WARNING_FORCE_NOT_CALIBRATED);
		force_calibrated = false;
	}
	else {
		force_calibrated = true;
		Notify::deactivate(Notify::messages::WARNING_FORCE_NOT_CALIBRATED);
	}

	// Initialize the limit compression and the Target compression
	protocol.data_register.limit_compression = System::Convert::ToByte(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_COMPRESSOR)[PaddleConfig::COMPRESSOR_MAX_FORCE]);
	if (protocol.data_register.limit_compression > 200) protocol.data_register.limit_compression = 200;
	if (protocol.data_register.limit_compression < 70) protocol.data_register.limit_compression = 70;

	
	protocol.data_register.target_compression = System::Convert::ToByte(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_COMPRESSOR)[PaddleConfig::COMPRESSOR_TARGET_FORCE]);
	if (protocol.data_register.target_compression >= protocol.data_register.limit_compression) protocol.data_register.target_compression = protocol.data_register.limit_compression;
	if (protocol.data_register.target_compression < 70) protocol.data_register.target_compression = 70;

	// Initializes the Holder data
	protocol.data_register.max_position = (unsigned short) System::Convert::ToInt32(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_COMPRESSOR)[PaddleConfig::COMPRESSOR_HOLDER_MAX_POSITION]);
	protocol.data_register.min_position = (unsigned short) System::Convert::ToInt32(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_COMPRESSOR)[PaddleConfig::COMPRESSOR_HOLDER_MIN_POSITION]);

}


/// <summary>
/// 
/// </summary>
/// 
/// This function is called by the Base class before to call the runningLoop() 
/// allowing the module to properly configure the device.
/// 
/// 
/// <param name=""></param>
/// <returns></returns>
bool PCB302::configurationLoop(void) {

	if (position_calibrated) {
		// Uploads the calibrated position parameters
		protocol.parameter_register.Kp = System::Convert::ToByte(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_COMPRESSOR)[PaddleConfig::COMPRESSOR_HOLDER_KP]);
		protocol.parameter_register.Op = (unsigned short)System::Convert::ToInt32(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_COMPRESSOR)[PaddleConfig::COMPRESSOR_HOLDER_OP]);
		writeParamRegister((int)ProtocolStructure::ParameterRegister::register_index::HOLDER_CALIB, protocol.parameter_register.encodeHolderCalibRegister());
	}
	
	
	if (force_calibrated) {
		// Uploads the force CALIB REGISTER parameters
		writeParamRegister((int)ProtocolStructure::ParameterRegister::register_index::COMPRESSION_CALIB, protocol.parameter_register.encodeCompressionCalibRegister());
	}
	

	return true;
}

/// <summary>
/// This function returns the index of the collimation format associated at the paddle.
/// 
/// The index of the collimation format should be in the range of 1:x where x 
/// should depend by the collimator number of available collimation format parameters.
/// 
///		NOTE: This module cannot check the index value that depends by other modules (the collimaotor).
/// 
/// 
/// </summary>
/// <param name="paddle_code">the code of the paddle to be investigated</param>
/// <returns>the collimation format or .-1 if the paddle is notn a valid paddle</returns>
System::Byte PCB302::getPaddleCollimationFormat(paddleCodes paddle_code) {

	try {
	
		switch ((paddleCodes)paddle_code) {
		case paddleCodes::PADDLE_PROSTHESIS: return System::Convert::ToByte(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_PROSTHESIS)[PaddleConfig::PADDLE_COLLIMATION]);
		case paddleCodes::PADDLE_BIOP2D:return System::Convert::ToByte(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_BIOP2D)[PaddleConfig::PADDLE_COLLIMATION]);
		case paddleCodes::PADDLE_BIOP3D:return System::Convert::ToByte(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_BIOP3D)[PaddleConfig::PADDLE_COLLIMATION]);
		case paddleCodes::PADDLE_TOMO:return System::Convert::ToByte(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_TOMO)[PaddleConfig::PADDLE_COLLIMATION]);
		case paddleCodes::PADDLE_24x30_CONTACT:return System::Convert::ToByte(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_24x30_CONTACT)[PaddleConfig::PADDLE_COLLIMATION]);
		case paddleCodes::PADDLE_18x24_C_CONTACT:return System::Convert::ToByte(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_18x24_C_CONTACT)[PaddleConfig::PADDLE_COLLIMATION]);
		case paddleCodes::PADDLE_18x24_L_CONTACT:return System::Convert::ToByte(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_18x24_L_CONTACT)[PaddleConfig::PADDLE_COLLIMATION]);
		case paddleCodes::PADDLE_18x24_R_CONTACT:return System::Convert::ToByte(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_18x24_R_CONTACT)[PaddleConfig::PADDLE_COLLIMATION]);
		case paddleCodes::PADDLE_10x24_CONTACT:return System::Convert::ToByte(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_10x24_CONTACT)[PaddleConfig::PADDLE_COLLIMATION]);
		case paddleCodes::PADDLE_9x21_MAG:return System::Convert::ToByte(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_9x21_MAG)[PaddleConfig::PADDLE_COLLIMATION]);
		case paddleCodes::PADDLE_9x9_MAG:return System::Convert::ToByte(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_9x9_MAG)[PaddleConfig::PADDLE_COLLIMATION]);
		case paddleCodes::PADDLE_D75_MAG:return System::Convert::ToByte(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_D75_MAG)[PaddleConfig::PADDLE_COLLIMATION]);
		}
	}
	catch (...) {
		return 0;
	}

	return  0;
}

/// <summary>
/// This function returns the paddle code from the paddle name
/// 
/// The Paddle name is a string name describing the Paddle.\n
/// The paddle name is used in the system for string protocols like the AWS protocol.
/// 
/// </summary>
/// <param name="tag">name of the paddle</param>
/// <returns>the paddle code or -1 if the no paddle is found</returns>
PCB302::paddleCodes PCB302::getPaddleCode(System::String^ tag) {
	for (int i = 0; i < (int) paddleCodes::PADDLE_LEN; i++) {
		if (tag == ((paddleCodes)i).ToString() ) return (paddleCodes) i;
	}

	return paddleCodes::PADDLE_NOT_DETECTED;
}



/// <summary>
/// This function returns the current detected paddle code
/// 
/// </summary>
/// <param name=""></param>
/// <returns>the current detected paddle code</returns>
PCB302::paddleCodes PCB302::getDetectedPaddleCode(void) {
	return detected_paddle;
}


void PCB302::getDetectedPaddleData(void) {
	static unsigned char current_tag = 255;
	if (protocol.status_register.paddle_tag == current_tag) return;
	current_tag = protocol.status_register.paddle_tag;

	// Find the data of the detected paddle exploring the configuration file
	if (System::Convert::ToByte(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_PROSTHESIS)[PaddleConfig::PADDLE_TAG]) == current_tag) {
		detected_paddle = paddleCodes::PADDLE_PROSTHESIS;
		detected_paddle_collimation_index = getPaddleCollimationFormat(detected_paddle);
		detected_paddle_offset = System::Convert::ToInt16(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_PROSTHESIS)[PaddleConfig::PADDLE_POSITION_OFFSET]);
		detected_paddle_weight = System::Convert::ToInt16(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_PROSTHESIS)[PaddleConfig::PADDLE_WEIGHT]);
		Notify::deactivate(Notify::messages::WARNING_MISSING_PADDLE);
		return;
	}

	if (System::Convert::ToByte(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_BIOP2D)[PaddleConfig::PADDLE_TAG]) == current_tag) {
		detected_paddle = paddleCodes::PADDLE_BIOP2D;
		detected_paddle_collimation_index = getPaddleCollimationFormat(detected_paddle);

		detected_paddle_offset = System::Convert::ToInt16(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_BIOP2D)[PaddleConfig::PADDLE_POSITION_OFFSET]);
		detected_paddle_weight = System::Convert::ToInt16(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_BIOP2D)[PaddleConfig::PADDLE_WEIGHT]);
		Notify::deactivate(Notify::messages::WARNING_MISSING_PADDLE);
		return;
	}

	if (System::Convert::ToByte(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_BIOP3D)[PaddleConfig::PADDLE_TAG]) == current_tag) {
		detected_paddle = paddleCodes::PADDLE_BIOP3D;
		detected_paddle_collimation_index = getPaddleCollimationFormat(detected_paddle);

		detected_paddle_offset = System::Convert::ToInt16(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_BIOP3D)[PaddleConfig::PADDLE_POSITION_OFFSET]);
		detected_paddle_weight = System::Convert::ToInt16(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_BIOP3D)[PaddleConfig::PADDLE_WEIGHT]);
		Notify::deactivate(Notify::messages::WARNING_MISSING_PADDLE);
		return;
	}

	if (System::Convert::ToByte(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_TOMO)[PaddleConfig::PADDLE_TAG]) == current_tag) {
		detected_paddle = paddleCodes::PADDLE_TOMO;
		detected_paddle_collimation_index = getPaddleCollimationFormat(detected_paddle);
		detected_paddle_offset = System::Convert::ToInt16(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_TOMO)[PaddleConfig::PADDLE_POSITION_OFFSET]);
		detected_paddle_weight = System::Convert::ToInt16(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_TOMO)[PaddleConfig::PADDLE_WEIGHT]);
		Notify::deactivate(Notify::messages::WARNING_MISSING_PADDLE);
		return;
	}

	if (System::Convert::ToByte(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_24x30_CONTACT)[PaddleConfig::PADDLE_TAG]) == current_tag) {
		detected_paddle = paddleCodes::PADDLE_24x30_CONTACT;
		detected_paddle_collimation_index = getPaddleCollimationFormat(detected_paddle);
		detected_paddle_offset = System::Convert::ToInt16(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_24x30_CONTACT)[PaddleConfig::PADDLE_POSITION_OFFSET]);
		detected_paddle_weight = System::Convert::ToInt16(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_24x30_CONTACT)[PaddleConfig::PADDLE_WEIGHT]);
		Notify::deactivate(Notify::messages::WARNING_MISSING_PADDLE);
		return;
	}

	if (System::Convert::ToByte(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_18x24_C_CONTACT)[PaddleConfig::PADDLE_TAG]) == current_tag) {
		detected_paddle = paddleCodes::PADDLE_18x24_C_CONTACT;
		detected_paddle_collimation_index = getPaddleCollimationFormat(detected_paddle);
		detected_paddle_offset = System::Convert::ToInt16(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_18x24_C_CONTACT)[PaddleConfig::PADDLE_POSITION_OFFSET]);
		detected_paddle_weight = System::Convert::ToInt16(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_18x24_C_CONTACT)[PaddleConfig::PADDLE_WEIGHT]);
		Notify::deactivate(Notify::messages::WARNING_MISSING_PADDLE);
		return;
	}

	if (System::Convert::ToByte(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_18x24_L_CONTACT)[PaddleConfig::PADDLE_TAG]) == current_tag) {
		detected_paddle = paddleCodes::PADDLE_18x24_L_CONTACT;
		detected_paddle_collimation_index = getPaddleCollimationFormat(detected_paddle);

		detected_paddle_offset = System::Convert::ToInt16(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_18x24_L_CONTACT)[PaddleConfig::PADDLE_POSITION_OFFSET]);
		detected_paddle_weight = System::Convert::ToInt16(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_18x24_L_CONTACT)[PaddleConfig::PADDLE_WEIGHT]);
		Notify::deactivate(Notify::messages::WARNING_MISSING_PADDLE);
		return;
	}

	if (System::Convert::ToByte(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_18x24_R_CONTACT)[PaddleConfig::PADDLE_TAG]) == current_tag) {
		detected_paddle = paddleCodes::PADDLE_18x24_R_CONTACT;
		detected_paddle_collimation_index = getPaddleCollimationFormat(detected_paddle);
		detected_paddle_offset = System::Convert::ToInt16(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_18x24_R_CONTACT)[PaddleConfig::PADDLE_POSITION_OFFSET]);
		detected_paddle_weight = System::Convert::ToInt16(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_18x24_R_CONTACT)[PaddleConfig::PADDLE_WEIGHT]);
		Notify::deactivate(Notify::messages::WARNING_MISSING_PADDLE);
		return;
	}

	if (System::Convert::ToByte(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_10x24_CONTACT)[PaddleConfig::PADDLE_TAG]) == current_tag) {
		detected_paddle = paddleCodes::PADDLE_10x24_CONTACT;
		detected_paddle_collimation_index = getPaddleCollimationFormat(detected_paddle);
		detected_paddle_offset = System::Convert::ToInt16(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_10x24_CONTACT)[PaddleConfig::PADDLE_POSITION_OFFSET]);
		detected_paddle_weight = System::Convert::ToInt16(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_10x24_CONTACT)[PaddleConfig::PADDLE_WEIGHT]);
		Notify::deactivate(Notify::messages::WARNING_MISSING_PADDLE);
		return;
	}

	if (System::Convert::ToByte(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_9x9_MAG)[PaddleConfig::PADDLE_TAG]) == current_tag) {
		detected_paddle = paddleCodes::PADDLE_9x9_MAG;
		detected_paddle_collimation_index = getPaddleCollimationFormat(detected_paddle);
		detected_paddle_offset = System::Convert::ToInt16(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_9x9_MAG)[PaddleConfig::PADDLE_POSITION_OFFSET]);
		detected_paddle_weight = System::Convert::ToInt16(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_9x9_MAG)[PaddleConfig::PADDLE_WEIGHT]);
		Notify::deactivate(Notify::messages::WARNING_MISSING_PADDLE);
		return;
	}

	if (System::Convert::ToByte(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_9x21_MAG)[PaddleConfig::PADDLE_TAG]) == current_tag) {
		detected_paddle = paddleCodes::PADDLE_9x21_MAG;
		detected_paddle_collimation_index = getPaddleCollimationFormat(detected_paddle);
		detected_paddle_offset = System::Convert::ToInt16(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_9x21_MAG)[PaddleConfig::PADDLE_POSITION_OFFSET]);
		detected_paddle_weight = System::Convert::ToInt16(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_9x21_MAG)[PaddleConfig::PADDLE_WEIGHT]);
		Notify::deactivate(Notify::messages::WARNING_MISSING_PADDLE);
		return;
	}

	if (System::Convert::ToByte(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_D75_MAG)[PaddleConfig::PADDLE_TAG]) == current_tag) {
		detected_paddle = paddleCodes::PADDLE_D75_MAG;
		detected_paddle_collimation_index = getPaddleCollimationFormat(detected_paddle);
		detected_paddle_offset = System::Convert::ToInt16(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_D75_MAG)[PaddleConfig::PADDLE_POSITION_OFFSET]);
		detected_paddle_weight = System::Convert::ToInt16(PaddleConfig::Configuration->getParam(PaddleConfig::PARAM_PADDLE_D75_MAG)[PaddleConfig::PADDLE_WEIGHT]);
		Notify::deactivate(Notify::messages::WARNING_MISSING_PADDLE);
		return;
	}

	detected_paddle = paddleCodes::PADDLE_NOT_DETECTED;
	detected_paddle_collimation_index = 0;
	detected_paddle_offset = 0;
	detected_paddle_weight = 0;
	Notify::activate(Notify::messages::WARNING_MISSING_PADDLE);
	return;

}


void PCB302::runningLoop(void) {
	static int count = 0;
	static bool commerr = false;

	// Test the communication status
	if (commerr != isCommunicationError()) {
		commerr = isCommunicationError();
		if (isCommunicationError()) {
			Notify::activate(Notify::messages::ERROR_PCB302_COMMUNICATION_ERROR);
		}
		else {
			Notify::deactivate(Notify::messages::ERROR_PCB302_COMMUNICATION_ERROR);
		}
	}

	//	Status Register
	protocol.status_register.decodeSystemRegister(readStatusRegister((unsigned char)ProtocolStructure::StatusRegister::register_index::SYSTEM_REGISTER));	
	std::this_thread::sleep_for(std::chrono::milliseconds(20));
	
	// Paddle Register
	protocol.status_register.decodePaddleRegister(readStatusRegister((unsigned char)ProtocolStructure::StatusRegister::register_index::PADDLE_REGISTER));
	std::this_thread::sleep_for(std::chrono::milliseconds(20));

	// The magnifier correction (if present)
	if (getMagnifierFactor() == 10) protocol.data_register.magnifier_offset = 0;
	else if (getMagnifierFactor() == 15) protocol.data_register.magnifier_offset = 150;
	else if (getMagnifierFactor() == 18) protocol.data_register.magnifier_offset = 180;
	else  protocol.data_register.magnifier_offset = 200;

	// Gets the detected paddle parameters: code, weight and offset
	getDetectedPaddleData();

	// Calculates the distance of the compression plate
	if (detected_paddle != paddleCodes::PADDLE_NOT_DETECTED) protocol.data_register.paddle_offset = detected_paddle_offset;
	else protocol.data_register.paddle_offset = 0;

	// Sets the current breast thickness and force
	if (protocol.status_register.compression_on) {

		breast_thickness = protocol.status_register.paddle_position - protocol.data_register.magnifier_offset - detected_paddle_offset;
		compression_force = protocol.status_register.paddle_force;
	}
	else {
		compression_force = 0;
		breast_thickness = 0;
	}
	
	
	// Sets the Paddle Weight data only if they are changed or if it is the first 
	// and if the Arm is properly initialized and calibrated
	static bool update_weight = true;

	// If the motor is not ready or not initialized doesn't update the current weight
	// In this case the update will be forced as soon as the motor returns to be ready and calibrated
	if ( (!ArmMotor::device->isReady()) || (!ArmMotor::device->isZeroOk()))  {
		update_weight = true;
	}
	else {
		// When the Arm motor is ready and calibrated (no running) 
		if (protocol.data_register.absolute_arm_angle != abs(ArmMotor::device->getCurrentPosition() / 100)) {
			protocol.data_register.absolute_arm_angle = abs(ArmMotor::device->getCurrentPosition() / 100);
			update_weight = true;
		}
		if (protocol.data_register.paddle_weight != detected_paddle_weight) {
			protocol.data_register.paddle_weight = detected_paddle_weight;
			update_weight = true;
		}
		if (update_weight) writeDataRegister((unsigned char)ProtocolStructure::DataRegister::register_index::PADDLE_WEIGHT, protocol.data_register.encodePaddleWeightRegister());
		update_weight = false;
	}
	
	writeDataRegister((int)ProtocolStructure::DataRegister::register_index::COMPRESSOR_LIMITS, protocol.data_register.encodeCompressorLimitsRegister());
	std::this_thread::sleep_for(std::chrono::milliseconds(20));
	
	writeDataRegister((int)ProtocolStructure::DataRegister::register_index::HOLDER_LIMITS, protocol.data_register.encodeHolderLimitsRegister());
	std::this_thread::sleep_for(std::chrono::milliseconds(20));

	return;
}

